### PR TITLE
Add event recorder into PipelineRun reconciler

### PR DESCRIPTION
**Pull request by JohnNiang**
### What this PR dose

Add event recorder into PipelineRun reconciler

### Why we need it

Please see https://github.com/kubesphere/ks-devops/pull/210#discussion_r699428922

### Special notes

image

/kind feature

### CODE REVIEW

File: controllers/jenkins/pipelinerun/pipelinerun_controller.go [hunk 1]
```diff
@@ -25,14 +25,33 @@ import (
...
+const (
+	Started        EventReason = "Stareted"
+	Updated        EventReason = "Updated"
+	TriggerFailed  EventReason = "TriggerFailed"
+	RetrieveFailed EventReason = "RetrieveFailed"
+)
+
+const (
+	Normal  EventType = "Normal"
+	Warning EventType = "Warning"
```
Comment (LinuxSuRen) [inline controllers/jenkins/pipelinerun/pipelinerun_controller.go:54]: Please consider reusing an existing one. See also

https://github.com/kubernetes/kubernetes/blob/0b4a793da2a2912393687367e0af2436612a9b8e/staging/src/k8s.io/api/core/v1/types.go#L5488

File: controllers/jenkins/pipelinerun/pipelinerun_controller.go [hunk 2]
```diff
@@ -25,14 +25,33 @@ import (
...
+type EventType string
+
+// EventReason is reason of event.
+type EventReason string
+
+const (
+	Started        EventReason = "Stareted"
+	Updated        EventReason = "Updated"
+	TriggerFailed  EventReason = "TriggerFailed"
+	RetrieveFailed EventReason = "RetrieveFailed"
```
Comment (LinuxSuRen) [inline controllers/jenkins/pipelinerun/pipelinerun_controller.go:49]: You don't need to convert `EventReason` to `string` if you don't create a type alias. Such as:

```
const (
  Updated string = "Updated"
)
```

See also:

* https://github.com/kubernetes/kubernetes/blob/0b4a793da2a2912393687367e0af2436612a9b8e/staging/src/k8s.io/api/core/v1/types.go#L5488
* https://github.com/kubesphere/ks-devops/blob/db195546d263e5fafa6829108bbca8ea982adc34/controllers/devopsproject/devopsproject_controller.go#L284

**Review by LinuxSuRen — CHANGES_REQUESTED** [review]

**Comment by JohnNiang** [comment]
When the parameterized PipelineRun has no parameters spec, the warning event appears as follow:

image

If we want get more details about the error,  https://github.com/jenkins-zh/jenkins-client need to be enhanced.

File: controllers/jenkins/pipelinerun/pipelinerun_controller.go (hunk 2 above)
Comment (JohnNiang) [reply controllers/jenkins/pipelinerun/pipelinerun_controller.go]: That's great.

File: controllers/jenkins/pipelinerun/pipelinerun_controller.go (hunk 1 above)
Comment (JohnNiang) [reply controllers/jenkins/pipelinerun/pipelinerun_controller.go]: Great.

**Review by LinuxSuRen — COMMENTED** [review]
/lgtm

**Review by LinuxSuRen — APPROVED** [review]

**Comment by JohnNiang** [comment]
Hi @LinuxSuRen , can this PR get ready for merging?

**Pull request merged**